File: src/slowql/rules/security/information.py
from __future__ import annotations

"""
Security Information rules.
"""



from slowql.core.models import Category, Dimension, Severity
from slowql.rules.base import PatternRule

__all__ = [
    'DatabaseVersionDisclosureRule',
    'SchemaInformationDisclosureRule',
    'TimingAttackPatternRule',
    'VerboseErrorMessageDisclosureRule',
]


class DatabaseVersionDisclosureRule(PatternRule):
    """Detects queries that expose database version information."""

    id = "SEC-INFO-001"
    name = "Database Version Disclosure"
    description = (
        "Detects queries that expose database version information, which aids attackers in finding "
        "version-specific vulnerabilities."
    )
    severity = Severity.LOW
    dimension = Dimension.SECURITY
    category = Category.SEC_DATA_EXPOSURE

    pattern = r'(?:@@VERSION|VERSION\(\)|SERVERPROPERTY\(\'ProductVersion\'\)|pg_version\(\)|BANNER|v\$version)'

    impact = (
        "Exposing database version helps attackers identify known vulnerabilities (CVEs) specific to that "
        "version. This information should not be accessible to application users."
    )
    fix_guidance = (
        "Never expose version info to end users. If needed for admin purposes, require authentication and "
        "log access. Return generic error messages without version details."
    )


class SchemaInformationDisclosureRule(PatternRule):
    """Detects queries accessing system catalog tables that expose schema information."""

    id = "SEC-INFO-002"
    name = "Schema Information Disclosure"
    description = (
        "Detects queries accessing system catalog tables that expose schema information to potential attackers."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.SECURITY
    category = Category.SEC_DATA_EXPOSURE

    pattern = r'\b(INFORMATION_SCHEMA|sys\.|pg_catalog|ALL_TABLES|USER_TABLES|DBA_TABLES|SHOW\s+TABLES|SHOW\s+COLUMNS|DESCRIBE|syscolumns|sysobjects)\b'

    impact = (
        "Schema enumeration reveals table names, column names, and relationships. Attackers use this for "
        "targeted SQL injection and privilege escalation. Should be restricted to DBAs only."
    )
    fix_guidance = (
        "Restrict access to system catalogs using database permissions. Don't expose schema info through "
        "application errors. Use views to hide underlying schema from application."
    )


class TimingAttackPatternRule(PatternRule):
    """Detects password/authentication queries without constant-time comparison."""

    id = "SEC-INFO-003"
    name = "Timing Attack Pattern"
    description = (
        "Detects password/authentication queries without constant-time comparison, enabling timing attacks."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.SECURITY
    category = Category.SEC_DATA_EXPOSURE

    pattern = r'\b(SLEEP|WAITFOR\s+DELAY|DBMS_LOCK\.SLEEP|PG_SLEEP)\b\s*\(\s*\d+\s*\)'

    impact = (
        "String comparison of passwords has variable timing based on match length. Attackers can infer "
        "password characters through timing analysis. Each character leak reduces brute-force complexity."
    )
    fix_guidance = (
        "Use constant-time comparison for password verification. Hash passwords and compare hashes. Add "
        "artificial delays to equalize timing. Use bcrypt/Argon2 which have built-in constant-time comparison."
    )


class VerboseErrorMessageDisclosureRule(PatternRule):
    """Detects error handling that may expose sensitive information."""

    id = "SEC-INFO-004"
    name = "Verbose Error Messages"
    description = (
        "Detects error handling that may expose sensitive information (stack traces, query text, schema details)."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.SECURITY
    category = Category.SEC_DATA_EXPOSURE

    pattern = r'\b(RAISERROR|THROW|SIGNAL)\b[^;]*\b(@@ERROR|ERROR_MESSAGE|SQLERRM|SQLSTATE)|\bCAST\s*\(\s*(?:@@VERSION|VERSION\(\)|BANNER)'

    impact = (
        "Error messages containing schema names, query fragments, or stack traces help attackers "
        "understand database structure and find injection points. Production errors should be generic."
    )
    fix_guidance = (
        "Return generic error messages to users ('An error occurred. Contact support.'). Log detailed "
        "errors server-side only. Never expose query text, object names, or internal errors to clients."
    )

File: slowql-rs/src/rules/security/information.rs
use crate::models::issue::Category;
use crate::models::{Dimension, Issue, Query, Severity};
use crate::rules::base::{RuleConfidence, Rule};
use once_cell::sync::Lazy;
use regex::Regex;

struct DatabaseVersionDisclosureRule;
static PAT_INFO_001: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)@@VERSION|VERSION\(\)|SERVERPROPERTY\('ProductVersion'\)|pg_version\(\)|BANNER|v\$version").unwrap()
});

impl Rule for DatabaseVersionDisclosureRule {
    fn id(&self) -> &'static str { "SEC-INFO-001" }
    fn name(&self) -> &'static str { "Database Version Disclosure" }
    fn severity(&self) -> Severity { Severity::Low }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDataExposure) }
    fn impact(&self) -> &'static str { "Exposing database version helps attackers identify known vulnerabilities (CVEs) specific to that version." }


    fn confidence(&self) -> RuleConfidence { RuleConfidence::Advisory }
    fn check(&self, query: &Query) -> Vec<Issue> { if query.source_context == "adhoc" || query.source_context.is_empty() { return Vec::new(); } PAT_INFO_001.find(&query.raw).map(|m| { let msg = format!("Database version disclosure: {}", m.as_str()); vec![self.build_issue(query, &msg, m.as_str())] }).unwrap_or_default() } }

struct SchemaInformationDisclosureRule;
static PAT_INFO_002: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\b(INFORMATION_SCHEMA|sys\.|pg_catalog|ALL_TABLES|USER_TABLES|DBA_TABLES|SHOW\s+TABLES|SHOW\s+COLUMNS|DESCRIBE|syscolumns|sysobjects)\b").unwrap()
});

impl Rule for SchemaInformationDisclosureRule {
    fn id(&self) -> &'static str { "SEC-INFO-002" }
    fn name(&self) -> &'static str { "Schema Information Disclosure" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDataExposure) }
    fn impact(&self) -> &'static str { "Schema enumeration reveals table names, column names, and relationships. Attackers use this for targeted SQL injection." }


    fn confidence(&self) -> RuleConfidence { RuleConfidence::Contextual }
    fn check(&self, query: &Query) -> Vec<Issue> {
        if query.source_context == "adhoc" || query.source_context.is_empty() { return Vec::new(); }
        if !PAT_INFO_002.is_match(&query.raw) { return Vec::new(); }
        // ORM introspection modules legitimately query schema metadata.
        // Skip when the file path indicates this is an adapter/introspection layer.
        if let Some(ref file) = query.location.file {
            let fl = file.to_lowercase();
            if fl.contains("introspection") || fl.contains("adapter")
                || fl.contains("backend") || fl.contains("connection_adapter")
                || fl.contains("schema_statements") {
                return Vec::new();
            }
        }
        // Legitimate uses: ORM introspection, migration tools, admin dashboards.
        // Only flag when there is evidence of dynamic enumeration or injection context:
        // concatenation, dynamic SQL, or UNION-based extraction pattern.
        let has_dynamic_signal = query.raw.contains("||")
            || query.raw.contains("CONCAT(")
            || query.raw.contains(" + ")
            || query.is_dynamic
            || query.raw_upper().contains("UNION")
            || query.raw_upper().contains("EXEC(")
            || query.raw_upper().contains("EXECUTE(");
        if !has_dynamic_signal { return Vec::new(); }
        PAT_INFO_002.find(&query.raw).map(|m| {
            let msg = format!("Schema information disclosure: {}", m.as_str());
            vec![self.build_issue(query, &msg, m.as_str())]
        }).unwrap_or_default()
    } }

struct TimingAttackPatternRule;
static PAT_INFO_003: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\b(SLEEP|WAITFOR\s+DELAY|DBMS_LOCK\.SLEEP|PG_SLEEP)\b\s*\(\s*\d+\s*\)").unwrap()
});

impl Rule for TimingAttackPatternRule {
    fn id(&self) -> &'static str { "SEC-INFO-003" }
    fn name(&self) -> &'static str { "Timing Attack Pattern" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDataExposure) }
    fn impact(&self) -> &'static str { "Attackers can infer password characters through timing analysis." }

    fn check(&self, query: &Query) -> Vec<Issue> {
        PAT_INFO_003.find(&query.raw).map(|m| {
            vec![self.build_issue(query, &format!("Timing attack pattern detected: {}", m.as_str()), m.as_str())]
        }).unwrap_or_default()
    }
}

struct VerboseErrorMessageDisclosureRule;
static PAT_INFO_004: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\b(RAISERROR|THROW|SIGNAL)\b[^;]*\b(@@ERROR|ERROR_MESSAGE|SQLERRM|SQLSTATE)|\bCAST\s*\(\s*(?:@@VERSION|VERSION\(\)|BANNER)").unwrap()
});

impl Rule for VerboseErrorMessageDisclosureRule {
    fn id(&self) -> &'static str { "SEC-INFO-004" }
    fn name(&self) -> &'static str { "Verbose Error Messages" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Security }
    fn category(&self) -> Option<Category> { Some(Category::SecDataExposure) }
    fn impact(&self) -> &'static str { "Error messages containing schema names or query fragments help attackers understand database structure." }

    fn check(&self, query: &Query) -> Vec<Issue> {
        PAT_INFO_004.find(&query.raw).map(|m| {
            vec![self.build_issue(query, &format!("Verbose error message disclosure: {}", m.as_str()), m.as_str())]
        }).unwrap_or_default()
    }
}

pub fn rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(DatabaseVersionDisclosureRule),
        Box::new(SchemaInformationDisclosureRule),
        Box::new(TimingAttackPatternRule),
        Box::new(VerboseErrorMessageDisclosureRule),
    ]
}
